### PR TITLE
Trees: add interleaved param groups to Def/Macro for parameter clause interleaving

### PR DESCRIPTION
**Pull request by kitbellew**
Also, add `Dialect.allowParamClauseInterleaving`. Based on #2820.

### CODE REVIEW

**Comment by kitbellew** [comment]
@Sporarium fyi

File: scalameta/trees/shared/src/main/scala/scala/meta/Trees.scala
Comment (kitbellew) [reply scalameta/trees/shared/src/main/scala/scala/meta/Trees.scala:505]: unfortunately, not every tree has an origin, and hence token positions.

another option is to add an explicit offset field to param clauses, but that might be a bit of a hack.

also, tparam is not a sequence but a single one, although we could just return the first one.

thoughts?

Comment (tgodzik) [reply scalameta/trees/shared/src/main/scala/scala/meta/Trees.scala:505]: > unfortunately, not every tree has an origin, and hence token positions.

That would be for transformed trees, no? Having `InterleavedDef` seems too much of a change and we would need to handle it everywhere we would handle Def, which might be a lot of additional cases.

Comment (kitbellew) [reply scalameta/trees/shared/src/main/scala/scala/meta/Trees.scala:505]: this applies to constructed (as opposed to parsed) trees. wouldn't be surprised that scalafix or any other ast-based application will not be able to use that.

Comment (tgodzik) [reply scalameta/trees/shared/src/main/scala/scala/meta/Trees.scala:505]: What we basically want here is having two fields replaced by one, I think that can't be handled by the current `repalacedField` no?

The other possibility is to have a way to match `InterleavedDef` with the standard Def unapply maybe? And could we make InterleavedDef extend Def? 

Or, we could have Def as a trait extended by OldDef and InterleavedDef?

Just throwing the ideas out there. Essentially an ideal situation would be that if someone enables clause itnerleaving then the same code that work without that dialect still works, otherwise it will all silently fail unless there are proper tests in place.

Comment (kitbellew) [reply scalameta/trees/shared/src/main/scala/scala/meta/Trees.scala:505]: @tgodzik figured out how to do it. a few intermediate PRs for you to review, before i update this one.

**Comment by kitbellew** [comment]
@tgodzik ready to go.

**Comment by tgodzik** [comment]
> @tgodzik ready to go.

Looks great! I will take a look a bit later this week, but are we ok releasing a new version before this gets merged just in case? `4.7.0` most likely, no?

**Comment by kitbellew** [comment]
> > @tgodzik ready to go.
> 
> Looks great! I will take a look a bit later this week, but are we ok releasing a new version before this gets merged just in case? `4.7.0` most likely, no?

if this is likely to be released, but we want to do it in next release, then i'd still like to add an intermediate change _in this release_. i will send it shortly.

**Comment by kitbellew** [comment]
> > > @tgodzik ready to go.
> > 
> > 
> > Looks great! I will take a look a bit later this week, but are we ok releasing a new version before this gets merged just in case? `4.7.0` most likely, no?
> 
> if this is likely to be released, but we want to do it in next release, then i'd still like to add an intermediate change _in this release_. i will send it shortly.

please take a look: #2987.

**Review by Sporarum — APPROVED** [review]
I believe this implements faithfully what is in the SIP
(given both the implementation and the tests)

Really nice work, thank you !

File: scalameta/dialects/shared/src/main/scala/scala/meta/Dialect.scala
Comment (kitbellew) [reply scalameta/dialects/shared/src/main/scala/scala/meta/Dialect.scala:161]: not supported in 2.11.

Comment (Sporarum) [reply scalameta/dialects/shared/src/main/scala/scala/meta/Dialect.scala:161]: Oh right, thanks

**Comment by tgodzik** [comment]
I think we should only merge this one after https://github.com/lampepfl/dotty/pull/14019 is finalized.

**Comment by Sporarum** [comment]
https://github.com/lampepfl/dotty/pull/14019 has been merged !

But it will of course not be usable until the next release of Scala

**Review by tgodzik — APPROVED** [review]
> [lampepfl/dotty#14019](https://github.com/lampepfl/dotty/pull/14019) has been merged !
> 
> But it will of course not be usable until the next release of Scala

It should be experimental in 3.3.x, right? We can merge it anyway in `Scala313` dialect I think, should safe.

**Comment by kitbellew** [comment]
> > [lampepfl/dotty#14019](https://github.com/lampepfl/dotty/pull/14019) has been merged !
> > But it will of course not be usable until the next release of Scala
> 
> It should be experimental in 3.3.x, right? We can merge it anyway in `Scala313` dialect I think, should safe.

meaning, instead of `Scala3Future`, add the flag to another dialect?

**Comment by tgodzik** [comment]
> > meaning, instead of `Scala3Future`, add the flag to another dialect?
> 
> Yeah, I would make it into the default Scala3 dialect, unless it could cause issues?

So instead of replying I edited your answer :face_with_spiral_eyes: Sorry about that! :| 

The answer:
I would make it into the default Scala3 dialect, unless it could cause issues?

**Comment by Sporarum** [comment]
> It should be experimental in 3.3.x, right?
As soon as 3.3.1

**Comment by kitbellew** [comment]
> > > meaning, instead of `Scala3Future`, add the flag to another dialect?
> > 
> > 
> > Yeah, I would make it into the default Scala3 dialect, unless it could cause issues?
> 
> So instead of replying I edited your answer 😵‍💫 Sorry about that! :|
> 
> The answer: I would make it into the default Scala3 dialect, unless it could cause issues?

Done. Added to 3.3 (which is what scala3 points to), along with fewer braces.

**Pull request merged**
